fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -412,15 +412,18 @@ public Void visitArrayExpr(ArrayExpr node, Scope scope) {
         @Override
         public Void visitMapExpr(MapExpr node, Scope scope) {
             if (!node.getChildren().isEmpty()) {
-                Type keyType = Type.NULL;
-                Type valueType = Type.NULL;
-                if (node.getKeyExpr() != null) {
-                    keyType = node.getKeyExpr().getType();
-                }
-                if (node.getValueExpr() != null) {
-                    valueType = node.getValueExpr().getType();
-                }
+                Type keyType = node.getKeyCommonType();
+                Type valueType = node.getValueCommonType();
                 node.setType(new MapType(keyType, valueType));
+                try {
+                    for (int i = 0; i < node.getChildren().size(); i++) {
+                        if (!node.getChildren().get(i).getType().matchesType(i % 2 == 0 ? keyType : valueType)) {
+                            node.castChild(i % 2 == 0 ? keyType : valueType, i);
+                        }
+                    }
+                } catch (AnalysisException e) {
+                    throw new SemanticException(e.getMessage());
+                }
```

Comment:
don‘t need add `cast` in analyzer

Reply:
it has to add cast to unify the type of keys or values.

Reply:
as array type does.

Reply:
I see, I meansing the cast should add in `implictCastRule`, not here. otherwise will lose cast logic when other people generate map expr in optimizer.